### DISCUSSION TITLE
Fail to read settings file

### DISCUSSION THREAD

**Discussion by nicoleMacias** [simnibs/simnibs#208] [7.Creating head models]
Hello,
I would like some inout in the following problem. I am currently following the affine registration troubleshooting steps. However, when I modify the settings file, the following error appears:

_[ simnibs ]INFO: simnibs version 4.0.1
[ simnibs ]INFO: charm run started: Mon Oct  2 11:04:29 2023
[ simnibs ]INFO: Reading settings from settings2.ini
[ simnibs ]ERROR: error when reading settings2.ini
[ simnibs ]CRITICAL: Uncaught exception
OSError: error when reading config file_

I have tried to change the name to 'settings' to see if that could be the issue, but it's not. 
It doesn't even let me initiate the atlas with the original settings file. 

Could it be I'm using the command wrong? 

I use : **charm patientNumber**(subID) **--initatlas --usesettings settings.ini**(or settings2 in the example)

Thank you!

**Comment by oulap** [comment]
Try passing the scan file in as well, like this:

charm patientNumber T1.nii.gz --initatlas --usesettings settings.ini

and see if that works. If that still gives you errors, can you paste the problematic settings file here?

**Reply by nicoleMacias** [reply]
Hi, thank you for the fast reply. Firstly, it asks to command --forcerun to rewrite the folder. And even when I did that, the same error (OSError: error when reading config file) appears. 

The settings are the following (I just modified affine_rotations and affine_vertical_shifts).

_# standard settings for charm.py
#  left-hand side should be in JSON synthax
#  int = 1
#  float = 1.2
#  bool = true/false (lower case)
#  str = "string" (double quotes)
#  list = [1, 2, 3]
#  dict = {"a": 1, "b": 2} (key must be string)

[preprocess]
denoise = false


# General settings for running samseg
[samseg]
# If threads is 0 the code uses all available threads, otherwise specify a positive integer number
threads = 8
atlas_name = "charm_atlas_mni"

# Custom intensity parameter file (default: empty)
# Note if a file is provided it should be in the same
# folder as the settings.ini file
gmm_parameter_file = ""

# Set to "mni" to use the "initmni" initialization below.
# Note this only works when a T1-weighted input is available.
# If no T1w scan is available, switch to "atlas".
init_type = "atlas"

# This is the default initialization for the segmentation
# when a T1-weighted input is available
[initmni]
# Translation scales for the optimizer
# (don't change this unless you know what you're doing default 1e-5)
translation_scale = -100
# Maximum number of iterations
max_iter = 300
# Shrink factor for the optimizer defined in 2^f.
# I.e., a shrink factor of 2 corresponds to downsampling
# by 2^2 = 4. Although 1 and 0 are valid inputs, less downsampling
# makes the initialization extremely slow, so not recommended.
shrink_factors = [2, 1, 0]
# A value for filling the background
bg_value = 0
# Smoothing factor for the intensities for optimization (in mm)
smoothing_factors = [4.0, 2.0, 0.0]
# Center-of-mass initialization, if set to zero the geometrical
# center is used instead.
center_of_mass = true
# Random sampling factor of the metric
samp_factor = 1.0


[initatlas]
# Scaling factors to search through as an initialization default: [0.85, 0.9, 0.95]
affine_scales = [[0.85, 0.85, 0.85], [0.9, 0.9, 0.85], [0.95, 0.95, 0.85]]
# Rotations (around LR axis, degrees) to search through default: [-7, -3.5, 0, 3.5, 7]
affine_rotations = [-5, -3, 0, 3, 5]
# Horizontal shifts (in mm) to search through default: [-20.0, -10.0, 0, 10.0, 20.0]
affine_horizontal_shifts = [-2.0, -1.0, 0, 1.0, 2.0]
# Vertical shifts to search through (in mm) default: [-10.0, 0.0, 10.0]
affine_vertical_shifts = [-10.0, 0.0, 10.0]
# Neck deformation bounds default: [-0.4, 0.1]
neck_search_bounds = [-0.3, 0.1]
# Downsampling target resolution (in mm) default: 2
downsampling_factor_affine = 2

[segment]
# Downsampling target resolution in mm default [2.0, 1.0]
downsampling_targets = [2.0, 1.0]
# Kernel size of bias field basis functions (in mm),
# smaller kernels are more flexible default: 70
bias_kernel_width = 70
# Smoothing factor for background mask in mm default: 4.0
background_mask_sigma = 4.0
# Threshold for masking background away default: 0.001
background_mask_threshold = 0.001
# Deformation penalty on the seg mesh  default: 0.1
mesh_stiffness = 0.1
# How to model Gaussian when multimodal data default: false
diagonal_covariances = false
# Downweighing factor for CSF when using T1+T2
csf_factor = 0.3

[surfaces]
# settings for creation of central GM surfaces
#If processes is 0, a separate process will be started for each surface
processes = 2
# Which central surfaces to reconstruct
surf = ["lh", "rh"]
# Which pial surfaces to reconstruct (has to be a subset of the central surfaces)
pial = ["lh", "rh"]
# see createCS in simnibs.segmentation.brain_surface
# for details of the below parameters
vdist = [1.0, 0.75]
voxsize_pbt = [0.5, 0.25]
voxsize_refinecs = [0.75, 0.5]
th_initial = 0.714
no_selfintersections = true
# Whether to ensure a minimal GM layer by filling in from the lh and rh central surfaces
fillin_gm_from_surf = true
# Whether to use the pial surfaces to reconstruct thin CSF structures
open_sulci_from_surf = true
# Prevent relabeling to GM in central brain regions: make ROI using tissues in labeling.nii.gz
exclusion_tissues_fillin_gm = {"left_cerebral_wm": 2,
                              "right_cerebral_wm": 41,
                              "stuff_to_exclude": [4, 10, 14, 16, 24, 28, 43, 49, 60]}
# Prevent relabeling to CSF in hippocampi and amygdalae: make ROI using tissues in labeling.nii.gz
exclusion_tissues_open_csf = [17, 18, 53, 54]

[mesh]
# Define thickness-dependent element sizes
# 	"range": upper and lower cut-offs for element sizes
# 	"slope": the smaller the finer the mesh
# 	"standard": applies to all tissues not defined otherwise
elem_sizes = {"standard": {"range": [1, 5], "slope": 1.0},
                     "1": {"range": [1, 7], "slope": 1.0},
                     "2": {"range": [1, 2], "slope": 1.0},
                     "5": {"range": [1, 10], "slope": 0.6}}
# Size of triangular kernel to smooth the size field. Can be 0.
smooth_size_field = 2
# Maximum size for the triangles of the outermost surface.
# 	Set to None to apply the same scaling as for the other elements.
skin_facet_size = 2.0
# Relation between thickness and maximally allowed distance of
# boundary facets to "true" label boundary
# 	"range": upper and lower cut-offs for the boundary facet distances
# 	"slope": the smaller the better the mesh follows the segmentation
facet_distances = {"standard": {"range": [0.1, 3], "slope": 0.5}}
# Whether to optimize the mesh using Lloyd. Takes some time
# but significanly improves element quality
optimize = false
# Whether to remove spikes from the mesh to create smoother mesh surfaces
remove_spikes = true
# Add outer surface to mesh using given tag. Set to false to disable.
skin_tag = 1005
# Hierarchy of surface tags (as list) that determines which triangle
# of a twin pair is kept. Set to false to use standard.
hierarchy = false
# Number of smoothing steps applied to the final surfaces. Can be 0.
smooth_steps = 10
# Number of addtional smoothing steps applied to the skin. Can be 0.
skin_care = 20_

**Reply by oulap** [reply]
Thanks, can you actually attach the settings file instead of copying the contents? 

I just want to see if I can reproduce the error.

**Reply by nicoleMacias** [reply]
Hi, yes sorry it didn't let me add a .ini. It's attached here. 
[settings2.txt](https://github.com/simnibs/simnibs/files/12784273/settings2.txt)

**Comment by oulap** [comment]
This works for me with the settings file you sent:

charm test_ernie ernie_T1.nii.gz --initatlas --usesettings settings2.ini

Can you fire up the simnibs python interpreter (by calling simnibs_python) and then run:

```
from simnibs import utils
settings = utils.settings_reader.read_ini('settings2.ini')
```

and see if that throws an error as well.

**Reply by nicoleMacias** [reply]
Hello,
I tried this aswell and it throws the same error. Most precisely, in the simnibs/utils/settings_reader.py line 38 in read_ini. 
It's the same OSError as before.

**Reply by oulap** [reply]
Hmm, and you're sure all the read and write permissions are okay for the settings2.ini file?

Can you try to run this in the same folder where the settings2.ini file is:

`simnibs_python -c "import configparser; cfg = configparser.ConfigParser(); fn = cfg.read('settings2.ini'); print(fn);"`

It should print: ['settings2.ini']

**Reply by nicoleMacias** [reply]
Hi! It does print it right.

**Reply by oulap** [reply]
But this one still throws an error?

```
from simnibs import utils
settings = utils.settings_reader.read_ini('settings2.ini')
```

That's strange as the error is triggered when the config parser returns an empty list.

**Reply by nicoleMacias** [reply]
Okay, in the same command window just after running the print, I did this again and it returns '[ simnibs ]INFO: Reading settings from settings2.ini'.

**Reply by oulap** [reply]
Okay so it seems to work.

Can you try to place the settings2.ini file to the same folder where you're running the charm command from, or pass the absolute path in the --usesettings flag? It might be that the settings file is just not found or couldn't be read for some other reason.

**Comment by nicoleMacias** [comment]
Okay, now it seems to be able to read the settings and just gives the the following error
_[ simnibs ]INFO: Starting affine registration and neck correction.
[ simnibs ]CRITICAL: Uncaught exception
IndexError: invalid axis: 0 (ndim = 0)_
Which I believe is maybe an error with the changes I made?

**Reply by nicoleMacias** [reply]
Turns out it returns the same for the original settings.

**Reply by oulap** [reply]
Do  you have singleton dimensions in your input scan? What is the shape of the buffer when you read the scan in?

**Reply by nicoleMacias** [reply]
I just check the .nii file with nibabel and supposedly there aren't any singleton dimensions (176, 256, 256). And same results with numpy. Is there any other way to check it? I'm new to these types of images.

**Comment by oulap** [comment]
Can you try to apply your settings file to the example data set we have available here:

https://github.com/simnibs/example-dataset/releases/download/v4.0/simnibs4_examples.zip

There's a T1 scan called ernie_T1.nii.gz under the /org-folder. 

Try running the charm call with that T1 as an input.

This is just to check if the issue is with your specific scan or the settings2.ini file.

**Reply by nicoleMacias** [reply]
I will do that next, but for the moment I can say that I tried running both, my file of settings and the original settings to initiate the atlas and the error was the same. As for the running the charm call, I suppose you mean to run it the conventional way? Or to run a charm from the beginning using --usesettings?

**Reply by oulap** [reply]
Right, that's true so it sounds like it's not something to do with the settings but maybe the scan itself.

Are you by any chance allowed to share the scan? It's a bit easier to debug the problem that way. In case you are and don't want to share the scan on github, you can also use our support email: 

support@simnibs.org

For the ernie test I was thinking about running charm with the --usesettings flag, but I presume that will be fine because you're getting the error with both the standard settings and the custom settings.

**Reply by nicoleMacias** [reply]
Thank you, I just sent the email to the support team and specified the discussion. I will be waiting for the response, thank you so much anyways!

**Reply by oulap** [reply]
The support email was unfortunately down. It's fixed now, but I never received the email. Could you try re-sending?

Thanks, and sorry for the hassle.

**Comment by nicoleMacias** [comment]
Of course, I resent it now!
